File: rust/src/shared/utils/phash.rs
use image::{ImageBuffer, Rgb, imageops};

/// Compute a 64-bit perceptual hash (pHash) for the given RGB24 pixels.
/// Algorithm (docs/models.md):
///   1. Resize to 32×32 grayscale
///   2. Apply 2D DCT-II
///   3. Take top-left 8×8 block (64 values)
///   4. Compute mean of those 64 values
///   5. Bit i = 1 if dct[i] > mean
///   6. Return as 16-char lowercase hex string
pub fn compute_phash(pixels: Vec<u8>, width: u32, height: u32) -> String {
    let img: ImageBuffer<Rgb<u8>, Vec<u8>> =
        ImageBuffer::from_raw(width, height, pixels)
            .expect("pixel buffer size does not match width × height × 3");

    // Resize to 32×32 and convert to grayscale float.
    let resized = imageops::resize(&img, 32, 32, imageops::FilterType::Triangle);
    let gray: Vec<f32> = resized
        .pixels()
        .map(|p| {
            let [r, g, b] = p.0;
            // Standard luminance weights.
            0.299 * r as f32 + 0.587 * g as f32 + 0.114 * b as f32
        })
        .collect();

    // 2D DCT-II (separable: rows then columns).
    let dct = dct2d(&gray, 32);

    // Top-left 8×8 block.
    let mut block = [0f32; 64];
    for row in 0..8usize {
        for col in 0..8usize {
            block[row * 8 + col] = dct[row * 32 + col];
        }
    }

    // Mean of 64 values.
    let mean: f32 = block.iter().sum::<f32>() / 64.0;

    // Pack 64 bits.
    let mut hash: u64 = 0;
    for (i, &val) in block.iter().enumerate() {
        if val > mean {
            hash |= 1u64 << i;
        }
    }

    format!("{:016x}", hash)
}

/// Separable 2D DCT-II on an N×N signal stored in row-major order.
/// Returns an N×N result in the same layout.
fn dct2d(signal: &[f32], n: usize) -> Vec<f32> {
    // Step 1: DCT-II along each row.
    let mut tmp = vec![0f32; n * n];
    for row in 0..n {
        dct1d(&signal[row * n..(row + 1) * n], &mut tmp[row * n..(row + 1) * n], n);
    }

    // Step 2: DCT-II along each column.
    let mut result = vec![0f32; n * n];
    let mut col_in = vec![0f32; n];
    let mut col_out = vec![0f32; n];
    for col in 0..n {
        for row in 0..n {
            col_in[row] = tmp[row * n + col];
        }
        dct1d(&col_in, &mut col_out, n);
        for row in 0..n {
            result[row * n + col] = col_out[row];
        }
    }

    result
}

/// 1D DCT-II: F[k] = sum_{x=0}^{N-1} f[x] * cos(pi*(2x+1)*k / (2N)).
/// Normalisation is omitted intentionally — only relative magnitudes matter for pHash.
fn dct1d(input: &[f32], output: &mut [f32], n: usize) {
    let scale = std::f32::consts::PI / (2.0 * n as f32);
    for k in 0..n {
        let mut sum = 0f32;
        for x in 0..n {
            sum += input[x] * ((2 * x + 1) as f32 * k as f32 * scale).cos();
        }
        output[k] = sum;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn solid_rgb(width: u32, height: u32, r: u8, g: u8, b: u8) -> Vec<u8> {
        let n = (width * height) as usize;
        let mut buf = Vec::with_capacity(n * 3);
        for _ in 0..n {
            buf.extend_from_slice(&[r, g, b]);
        }
        buf
    }

    #[test]
    fn phash_length_is_16() {
        let pixels = solid_rgb(64, 64, 128, 64, 32);
        let hash = compute_phash(pixels, 64, 64);
        assert_eq!(hash.len(), 16, "pHash must be a 16-char hex string, got: {hash}");
    }

    #[test]
    fn phash_is_valid_hex() {
        let pixels = solid_rgb(64, 64, 200, 100, 50);
        let hash = compute_phash(pixels, 64, 64);
        assert!(
            hash.chars().all(|c| c.is_ascii_hexdigit()),
            "pHash must contain only hex digits, got: {hash}"
        );
    }

    #[test]
    fn identical_images_have_identical_hash() {
        let pixels = solid_rgb(128, 128, 77, 88, 99);
        let h1 = compute_phash(pixels.clone(), 128, 128);
        let h2 = compute_phash(pixels, 128, 128);
        assert_eq!(h1, h2, "identical images must produce identical pHash");
    }

    #[test]
    fn hamming_distance_of_same_image_is_zero() {
        let pixels = solid_rgb(64, 64, 10, 20, 30);
        let h1 = u64::from_str_radix(&compute_phash(pixels.clone(), 64, 64), 16).unwrap();
        let h2 = u64::from_str_radix(&compute_phash(pixels, 64, 64), 16).unwrap();
        assert_eq!((h1 ^ h2).count_ones(), 0);
    }
}

File: lib/core/inference/phash.dart
import 'dart:math' as math;
import 'dart:typed_data';

import 'package:ai_gallery/core/inference/image_tensor_utils.dart';
import 'package:image/image.dart' as img;

String computePhashDart(Uint8List pixels, int width, int height) {
  final source = rgbImageFromBytes(pixels, width, height);
  final resized = img.copyResize(
    source,
    width: 32,
    height: 32,
    interpolation: img.Interpolation.linear,
  );

  final gray = List<double>.filled(32 * 32, 0);
  for (var y = 0; y < 32; y++) {
    for (var x = 0; x < 32; x++) {
      final p = resized.getPixel(x, y);
      gray[y * 32 + x] = (0.299 * p.r + 0.587 * p.g + 0.114 * p.b).toDouble();
    }
  }

  final dct = _dct2d(gray, 32);
  final block = List<double>.filled(64, 0);
  for (var row = 0; row < 8; row++) {
    for (var col = 0; col < 8; col++) {
      block[row * 8 + col] = dct[row * 32 + col];
    }
  }

  final mean = block.reduce((a, b) => a + b) / 64.0;
  var hash = BigInt.zero;
  for (var i = 0; i < block.length; i++) {
    if (block[i] > mean) {
      hash |= BigInt.one << i;
    }
  }
  return hash.toRadixString(16).padLeft(16, '0');
}

List<double> _dct2d(List<double> signal, int n) {
  final tmp = List<double>.filled(n * n, 0);
  for (var row = 0; row < n; row++) {
    final input = signal.sublist(row * n, (row + 1) * n);
    final output = _dct1d(input, n);
    tmp.setRange(row * n, (row + 1) * n, output);
  }

  final result = List<double>.filled(n * n, 0);
  for (var col = 0; col < n; col++) {
    final input = List<double>.filled(n, 0);
    for (var row = 0; row < n; row++) {
      input[row] = tmp[row * n + col];
    }
    final output = _dct1d(input, n);
    for (var row = 0; row < n; row++) {
      result[row * n + col] = output[row];
    }
  }
  return result;
}

List<double> _dct1d(List<double> input, int n) {
  final output = List<double>.filled(n, 0);
  final scale = math.pi / (2.0 * n);
  for (var k = 0; k < n; k++) {
    var sum = 0.0;
    for (var x = 0; x < n; x++) {
      sum += input[x] * math.cos((2 * x + 1) * k * scale);
    }
    output[k] = sum;
  }
  return output;
}
